design-proposals/klip-25-removal-of-with-key-syntax.md

```diff
@@ -0,0 +1,163 @@
+
+## Motivation and background
+
+The `WITH(KEY)` syntax allows a user to provide an alias for the `ROWKEY` system column in ksqlDB:
+
+```sql
+-- old skool syntax
+CREATE TABLE INPUT (ID INT, NAME STRING) WITH ('KEY'='ID', 'KAFKA_TOPIC'='input', ...);
+```
+
+The above statement defines a table with an implicit STRING primary key column called `ROWKEY` and
+two other columns called `ID` and `NAME`.  The presence of the `WITH('KEY'='ID')` tells ksqlDB that
+the `ID` column is essentially an aliases for `ROWKEY` as it contains the exact same data.
+
+### Aliasing
+
+Allowing users to provide an alias for the system column `ROWKEY` allowed users to provide a more
+domain driven name, making queries more expressive:
+
+```sql
+-- Given a stream of users where the key contained a region code:
+
+CREATE TABLE USERS_BY_REGION AS
+   SELECT COUNT(*) FROM USERS GROUP BY ROWKEY;
+
+-- is less expressive than --
+
+CREATE TABLE USERS_BY_REGION AS
+   SELECT REGION, COUNT(*) FROM USERS GROUP BY REGION;
+```
+
+However, [a feature removing the restriction that key columns must be called `ROWKEY` is about to be
+enabled](https://github.com/confluentinc/ksql/pull/5093). Once it is, the aliasing that `WITH(KEY)`
+provides will be redundant, as the same can be achieved much more intuitively, and without the need
+to duplicate the key data into a value field, using the new syntax. For example:
+
+```sql
+-- old skool syntax:
+CREATE TABLE USERS (ROWKEY BIGINT PRIMARY KEY, NAME STRING, ID BIGINT) WITH ('KEY'='ID', ...);
+
+-- new skool syntax:
+CREATE TABLE USERS (ID BIGINT PRIMARY KEY, NAME STRING) WITH (...);
+```
+
+### Optimisations
+
+This aliasing allows ksqlDB to optimize certain queries by avoiding unnecessary repartition steps,
+should the 'key field' be used in a `PARTITION BY`, `GROUP BY` or `JOIN ON` clause. For example:
+
+```sql
+-- given:
+CREATE TABLE INPUT (ID INT, NAME STRING) WITH ('KEY'='ID', 'KAFKA_TOPIC'='input', ...);
+
+-- no repartition necessary as `ID` is equivalent to `ROWKEY`
+SELECT ID, COUNT() FROM INPUT GROUP BY ID;
+```
+
+KsqlDB can avoid an unnecessary repartition step in the above statement as `ID` is equivalent to
+`ROWKEY`, and grouping by `ROWKEY` is already correctly partitioned.
+
+It's not just streams and tables created by `CREATE STREAM` and `CREATE TABLE` statements that can
+track these `key-fields`. KsqDB also tries to keep track of the key field in derived sources
+too, for example:
+
+```sql
+CREATE TABLE FOO AS ID, COUNT() FROM BAR GROUP BY ID;
+-- Foo has key-field 'ID'
+```
+
+In the above, KsqlDB will detect that `FOO``s `ID` column is equivalent to `ROWKEY`. However, it can
+only handle a single field. If there are multiple, only the first is tracked, for example:
+
+```sql
+CREATE TABLE FOO AS ID AS ID1, ID AS ID2, COUNT() FROM BAR GROUP BY ID;
+-- Foo has key-field 'ID1'
+-- `ID2` is not a key field, even though its a duplicate of `ID1`
```

Comment:
Orthogonal to this KLIP: why is this not allowed? Seems weird to me and I believe it would be ok with standard SQL and I see. no reason (given the current improvements for key handling) to not allow this? Are there any plans to fix it (I guess it won't need a KLIP; seems to be a follow up bug fix / improvement after the current work on keys is done)?

Reply:
Totally - it would be possible for KSQL to track the full set of 'columns that are just an alias of the key' for each source. This would allow us to avoid unnecessary repartitions.  

However, the ROI on such work may be limited in the short to mid term compared to other work. Especially considering that users will soon be able to name their key columns as they like, and hence could just join on those columns.  In the new model, how often will people be creating duplicates of the key column elsewhere? Hard to quantify, but _probably_ not that much.  Also, once support for multiple key columns are added, the complexity of supporting this goes up, while the usefulness / ROI may well go down.

One could argue that removing the current partial / buggy implementation of this is a step backwards. However, I would argue that without the investment to fix the bugs and extend to a full implementation, the current implementation has a _negative_ impact on functionality (and code), overall and is better removed.

Regardless of what we do with the current impl, the syntax should definitely be removed as any internal tracking of key aliases should be just that: internal.

Reply:
Thanks for the details!